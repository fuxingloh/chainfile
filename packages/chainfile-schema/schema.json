{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "allOf": [
    {
      "$ref": "#/definitions/Chainfile"
    }
  ],
  "definitions": {
    "Chainfile": {
      "type": "object",
      "properties": {
        "$schema": {
          "type": "string"
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        },
        "caip2": {
          "type": "string",
          "pattern": "^[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}$",
          "description": "CAIP-2 Chain ID of the blockchain network this chainfile is for."
        },
        "name": {
          "type": "string",
          "description": "Describes this chainfile.",
          "minLength": 1,
          "maxLength": 128
        },
        "params": {
          "type": "object",
          "maxProperties": 30,
          "patternProperties": {
            "^[a-z]+(_[a-z0-9]+)*$": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamOptions"
                }
              ]
            }
          },
          "additionalProperties": false
        },
        "volumes": {
          "type": "object",
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Volume"
            }
          },
          "additionalProperties": false
        },
        "containers": {
          "type": "object",
          "minProperties": 1,
          "maxProperties": 6,
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Container"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["caip2", "name", "containers"],
      "additionalProperties": false
    },
    "ParamOptions": {
      "type": "object",
      "properties": {
        "secret": {
          "type": "boolean",
          "description": "Mark the value as a secret"
        },
        "default": {
          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "object",
              "properties": {
                "random": {
                  "type": "object",
                  "description": "Generate a random value for orchestration that supports it.",
                  "properties": {
                    "bytes": {
                      "type": "number",
                      "minimum": 1,
                      "maximum": 256,
                      "multipleOf": 1
                    },
                    "encoding": {
                      "enum": ["hex", "base64", "base64url"]
                    }
                  },
                  "required": ["bytes", "encoding"],
                  "additionalProperties": false
                }
              },
              "additionalProperties": false
            }
          ]
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        }
      },
      "additionalProperties": false
    },
    "ParamReference": {
      "type": "object",
      "properties": {
        "$param": {
          "type": "string",
          "description": "Name of the value to reference in the params section.",
          "pattern": "^[a-z]+(_[a-z0-9]+)*$"
        }
      },
      "required": ["$param"],
      "additionalProperties": false
    },
    "Volume": {
      "type": "object",
      "properties": {
        "type": {
          "enum": ["persistent", "ephemeral"]
        },
        "size": {
          "type": "string",
          "pattern": "^[1-9][0-9]*[MGT]i$"
        },
        "expansion": {
          "type": "object",
          "description": "Optional growth rate for volumes for evolving storage needs. [(size) + (rate * time)]",
          "properties": {
            "startFrom": {
              "type": "string",
              "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
              "description": "Formatted as YYYY-MM-DD to denote the date the growth rate starts."
            },
            "monthlyRate": {
              "type": "string",
              "pattern": "^[1-9][0-9]*[MGT]i$",
              "description": "The monthly rate at which the volume grows."
            }
          },
          "required": ["startFrom", "monthlyRate"],
          "additionalProperties": false
        }
      },
      "required": ["type", "size"],
      "additionalProperties": false
    },
    "VolumeMount": {
      "type": "object",
      "properties": {
        "volume": {
          "type": "string",
          "pattern": "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$",
          "description": "Name of the volume to mount, defined in the volumes section."
        },
        "mountPath": {
          "type": "string"
        },
        "subPath": {
          "type": "string"
        }
      },
      "required": ["volume", "mountPath"],
      "additionalProperties": false
    },
    "Container": {
      "type": "object",
      "properties": {
        "image": {
          "type": "string"
        },
        "tag": {
          "description": "Tag of the container image.",
          "oneOf": [
            {
              "type": "string"
            },
            {
              "$ref": "#/definitions/ParamReference"
            }
          ]
        },
        "source": {
          "type": "string",
          "description": "Source of the container image.",
          "format": "uri",
          "minLength": 1,
          "maxLength": 256
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        },
        "endpoints": {
          "type": "object",
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Endpoint"
            }
          },
          "additionalProperties": false
        },
        "resources": {
          "type": "object",
          "description": "For placement of the node in a container cluster.",
          "properties": {
            "cpu": {
              "type": "number",
              "description": "CPU in vCPU (Virtual CPU Core, Relative). In multiples of 0.25, where 0.25 is 1/4 of a core.",
              "minimum": 0.25,
              "maximum": 16,
              "multipleOf": 0.25
            },
            "memory": {
              "type": "number",
              "description": "Memory in MiB (Mebibyte, Absolute). You are guaranteed to have this much memory available to your container.",
              "minimum": 64,
              "maximum": 122880,
              "multipleOf": 64
            }
          },
          "required": ["cpu", "memory"],
          "additionalProperties": false
        },
        "environment": {
          "type": "object",
          "maxProperties": 30,
          "patternProperties": {
            "^[a-zA-Z_]+[a-zA-Z0-9_]*$": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "additionalProperties": false
        },
        "command": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "mounts": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/VolumeMount"
          }
        }
      },
      "required": ["image", "tag", "source", "resources"],
      "additionalProperties": false
    },
    "Endpoint": {
      "type": "object",
      "description": "Extension to ports, endpoints are used to expose the container to the outside world with a specific protocol for interfacing.",
      "properties": {
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        }
      },
      "oneOf": [
        {
          "$ref": "#/definitions/EndpointPort"
        },
        {
          "$ref": "#/definitions/EndpointHttpRest"
        },
        {
          "$ref": "#/definitions/EndpointHttpJsonRpc"
        }
      ]
    },
    "EndpointPort": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        }
      },
      "required": ["port"],
      "additionalProperties": false
    },
    "EndpointHttpRest": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP REST", "HTTPS REST"]
        },
        "authorization": {
          "$ref": "#/definitions/EndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "EndpointHttpRestProbe": {
      "type": "object",
      "properties": {
        "method": {
          "enum": ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS", "TRACE"]
        },
        "path": {
          "type": "string"
        },
        "body": {
          "type": "object"
        },
        "match": {
          "type": "object",
          "properties": {
            "status": {
              "oneOf": [
                {
                  "type": "number",
                  "minimum": 100,
                  "maximum": 599,
                  "multipleOf": 1
                },
                {
                  "type": "array",
                  "uniqueItems": true,
                  "items": {
                    "type": "number",
                    "minimum": 100,
                    "maximum": 599,
                    "multipleOf": 1
                  }
                }
              ]
            },
            "body": {
              "$ref": "http://json-schema.org/draft-07/schema#"
            }
          },
          "required": ["status"]
        }
      },
      "required": ["method", "path", "match"],
      "additionalProperties": false
    },
    "EndpointHttpJsonRpc": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP JSON-RPC 1.0", "HTTPS JSON-RPC 1.0", "HTTP JSON-RPC 2.0", "HTTPS JSON-RPC 2.0"]
        },
        "path": {
          "type": "string"
        },
        "authorization": {
          "$ref": "#/definitions/EndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "EndpointHttpJsonRpcProbe": {
      "type": "object",
      "properties": {
        "method": {
          "type": "string"
        },
        "params": {
          "oneOf": [
            {
              "type": "array"
            },
            {
              "type": "object"
            }
          ]
        },
        "match": {
          "type": "object",
          "properties": {
            "result": {
              "$ref": "http://json-schema.org/draft-07/schema"
            }
          },
          "required": ["result"]
        }
      },
      "required": ["method", "params", "match"],
      "additionalProperties": false
    },
    "EndpointHttpAuthorization": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBasic"
            },
            "username": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            },
            "password": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "required": ["type", "username", "password"],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBearer"
            },
            "token": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "required": ["type", "token"],
          "additionalProperties": false
        }
      ]
    }
  }
}
