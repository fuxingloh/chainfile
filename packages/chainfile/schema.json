{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://chainfile.org/schema.json",
  "title": "Chainfile",
  "description": "Chainfile defines the instructions on how to package complex blockchain nodes into a Container that can be easily deployed and managed on Container-capable platforms such as Kubernetes, Compose, and ECS.",
  "type": "object",
  "properties": {
    "$schema": {
      "type": "string"
    },
    "id": {
      "type": "string",
      "pattern": "^[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}/.+$",
      "description": "CAIP-2 Extended ID that uniquely identifies this chainfile."
    },
    "caip2": {
      "type": "string",
      "pattern": "^[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}$",
      "description": "CAIP-2 Chain ID of the blockchain network this chainfile is for."
    },
    "name": {
      "type": "string",
      "description": "Describes this chainfile.",
      "minLength": 1,
      "maxLength": 128
    },
    "env": {
      "$ref": "#/definitions/Env"
    },
    "containers": {
      "type": "object",
      "minProperties": 1,
      "maxProperties": 6,
      "patternProperties": {
        "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
          "$ref": "#/definitions/Container"
        }
      },
      "additionalProperties": false
    }
  },
  "required": ["id", "caip2", "name", "containers"],
  "additionalProperties": false,
  "definitions": {
    "Env": {
      "type": "object",
      "maxProperties": 30,
      "patternProperties": {
        "^[a-zA-Z_]+[a-zA-Z0-9_]*$": {
          "$ref": "#/definitions/EnvFactory"
        }
      },
      "additionalProperties": false
    },
    "EnvFactory": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "RandomBytes"
            },
            "length": {
              "type": "number",
              "minimum": 1,
              "maximum": 256,
              "multipleOf": 1
            },
            "encoding": {
              "enum": ["hex", "base64", "base64url"]
            }
          },
          "required": ["type", "length", "encoding"],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "Value"
            },
            "value": {
              "type": "string"
            }
          },
          "required": ["type", "value"],
          "additionalProperties": false
        }
      ]
    },
    "EnvReference": {
      "type": "object",
      "properties": {
        "key": {
          "type": "string"
        }
      },
      "required": ["key"],
      "additionalProperties": false
    },
    "Container": {
      "type": "object",
      "properties": {
        "image": {
          "type": "string"
        },
        "source": {
          "type": "string",
          "description": "Source of the container image.",
          "format": "uri"
        },
        "endpoints": {
          "type": "object",
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/ContainerEndpoint"
            }
          },
          "additionalProperties": false
        },
        "resources": {
          "type": "object",
          "description": "For placement of the node in a container cluster.",
          "properties": {
            "cpu": {
              "type": "number",
              "description": "CPU in vCPU (Virtual CPU Core, Relative). In multiples of 0.25, where 0.25 is 1/4 of a core.",
              "minimum": 0.25,
              "maximum": 16,
              "multipleOf": 0.25
            },
            "memory": {
              "type": "number",
              "description": "Memory in MiB (Mebibyte, Absolute). You are guaranteed to have this much memory available to your container.",
              "minimum": 64,
              "maximum": 122880,
              "multipleOf": 64
            }
          },
          "required": ["cpu", "memory"],
          "additionalProperties": false
        },
        "environment": {
          "type": "object",
          "maxProperties": 30,
          "patternProperties": {
            "^[a-zA-Z_]+[a-zA-Z0-9_]*$": {
              "oneOf": [{ "type": "string" }, { "$ref": "#/definitions/EnvReference" }]
            }
          },
          "additionalProperties": false
        },
        "command": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "volumes": {
          "type": "object",
          "properties": {
            "persistent": {
              "$ref": "#/definitions/ContainerVolume"
            },
            "ephemeral": {
              "$ref": "#/definitions/ContainerVolume"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["image", "source", "endpoints", "resources"],
      "additionalProperties": false
    },
    "ContainerEndpoint": {
      "type": "object",
      "description": "Extension to ports, endpoints are used to expose the container to the outside world with a specific protocol for interfacing.",
      "oneOf": [
        {
          "$ref": "#/definitions/ContainerEndpointPort"
        },
        {
          "$ref": "#/definitions/ContainerEndpointHttpRest"
        },
        {
          "$ref": "#/definitions/ContainerEndpointHttpJsonRpc"
        }
      ]
    },
    "ContainerEndpointPort": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        }
      },
      "required": ["port"],
      "additionalProperties": false
    },
    "ContainerEndpointHttpRest": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP REST", "HTTPS REST"]
        },
        "authorization": {
          "$ref": "#/definitions/ContainerEndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/ContainerEndpointHttpRestProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/ContainerEndpointHttpRestProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/ContainerEndpointHttpRestProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "ContainerEndpointHttpRestProbe": {
      "type": "object",
      "properties": {
        "method": {
          "enum": ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS", "TRACE"]
        },
        "path": {
          "type": "string"
        },
        "body": {
          "type": "object"
        },
        "match": {
          "type": "object",
          "properties": {
            "status": {
              "oneOf": [
                {
                  "type": "number",
                  "minimum": 100,
                  "maximum": 599,
                  "multipleOf": 1
                },
                {
                  "type": "array",
                  "uniqueItems": true,
                  "items": {
                    "type": "number",
                    "minimum": 100,
                    "maximum": 599,
                    "multipleOf": 1
                  }
                }
              ]
            },
            "body": {
              "$ref": "http://json-schema.org/draft-07/schema#"
            }
          },
          "required": ["status"]
        }
      },
      "required": ["method", "path", "match"],
      "additionalProperties": false
    },
    "ContainerEndpointHttpJsonRpc": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP JSON-RPC 1.0", "HTTPS JSON-RPC 1.0", "HTTP JSON-RPC 2.0", "HTTPS JSON-RPC 2.0"]
        },
        "path": {
          "type": "string"
        },
        "authorization": {
          "$ref": "#/definitions/ContainerEndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/ContainerEndpointHttpJsonRpcProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/ContainerEndpointHttpJsonRpcProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/ContainerEndpointHttpJsonRpcProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "ContainerEndpointHttpJsonRpcProbe": {
      "type": "object",
      "properties": {
        "method": {
          "type": "string"
        },
        "params": {
          "oneOf": [
            {
              "type": "array"
            },
            {
              "type": "object"
            }
          ]
        },
        "match": {
          "type": "object",
          "properties": {
            "result": {
              "$ref": "http://json-schema.org/draft-07/schema"
            }
          },
          "required": ["result"]
        }
      },
      "required": ["method", "params", "match"],
      "additionalProperties": false
    },
    "ContainerEndpointHttpAuthorization": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBasic"
            },
            "username": {
              "oneOf": [{ "type": "string" }, { "$ref": "#/definitions/EnvReference" }]
            },
            "password": {
              "oneOf": [{ "type": "string" }, { "$ref": "#/definitions/EnvReference" }]
            }
          },
          "required": ["type", "username", "password"],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBearer"
            },
            "token": {
              "oneOf": [{ "type": "string" }, { "$ref": "#/definitions/EnvReference" }]
            }
          },
          "required": ["type", "token"],
          "additionalProperties": false
        }
      ]
    },
    "ContainerVolume": {
      "type": "object",
      "properties": {
        "paths": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "size": {
          "oneOf": [
            {
              "type": "string",
              "pattern": "^[1-9][0-9]*([MGT])$"
            },
            {
              "type": "object",
              "properties": {
                "initial": {
                  "type": "string",
                  "pattern": "^[1-9][0-9]*([MGT])$"
                },
                "from": {
                  "type": "string",
                  "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
                  "description": "YYYY-MM-DD"
                },
                "growth": {
                  "type": "string",
                  "pattern": "^[1-9][0-9]*([MGT])$"
                },
                "rate": {
                  "enum": ["daily", "weekly", "monthly", "yearly"]
                }
              },
              "required": ["initial", "from", "growth", "rate"],
              "additionalProperties": false
            }
          ]
        }
      },
      "required": ["paths", "size"],
      "additionalProperties": false
    }
  }
}
