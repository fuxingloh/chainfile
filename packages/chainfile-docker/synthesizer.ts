import { randomBytes } from 'node:crypto';

import { ChainfileDefinition, Container } from 'chainfile';
import { synthEnv } from 'chainfile/environment';
import { validate } from 'chainfile/schema';
import yaml from 'js-yaml';

import { version } from './package.json';

/**
 * Synthesize a ChainfileDefinition into `docker.*.yml` & `.env` files.
 */
export class Synthesizer {
  private readonly definition: ChainfileDefinition;
  /**
   * A random deploymentId is generated for each synthesizer to uniquely the lifetime of a deployment.
   */
  public readonly deploymentId: string = randomBytes(8).toString('hex');

  constructor(definition: ChainfileDefinition | any) {
    validate(definition);
    this.definition = definition;
  }

  public synthEnv(): string {
    return Object.entries({
      ...synthEnv(this.definition.environment ?? {}),
      CHAINFILE_DEPLOYMENT_ID: this.deploymentId,
    })
      .map(([key, value]) => `${key}=${value}`)
      .join('\n');
  }

  public synthCompose(): string {
    return [
      `# Generated by chainfile-docker, do not edit manually.`,
      `# Chainfile Version: ${version}`,
      `# Definition ID: ${this.definition.id}`,
      `# Definition Name: ${this.definition.name}`,
      `# Definition CAIP-2: ${this.definition.caip2}`,
      '',
      yaml.dump(
        {
          name: this.definition.id.replaceAll(/[^a-z0-9_-]/g, '_'),
          services: {
            ...this.createChainfileAgent(),
            ...this.createContainers(),
          },
          networks: {
            chainfile: {},
          },
          volumes: {
            chainfile: {},
          },
        },
        {
          lineWidth: 120,
        },
      ),
    ].join('\n');
  }

  private createChainfileAgent(): Record<'chainfile-agent', object> {
    const definitionJson = JSON.stringify({
      ...this.definition,
      $schema: undefined,
    }).replaceAll('$', '$$$');

    const EnvironmentMapping = Object.keys(this.definition.environment ?? {}).reduce(
      (env, key) => {
        env[`CHAINFILE_ENVIRONMENT_${key}`] = `$\{${key}}`;
        return env;
      },
      {} as Record<string, string>,
    );

    return {
      'chainfile-agent': {
        container_name: 'chainfile-agent-${CHAINFILE_DEPLOYMENT_ID}',
        image: `ghcr.io/vetumorg/chainfile-agent:${version}`,
        ports: ['0:1503'],
        environment: {
          // Docker compose automatically evaluate environment literals here
          CHAINFILE_DEFINITION_JSON: definitionJson,
          CHAINFILE_DEPLOYMENT_ID: '${CHAINFILE_DEPLOYMENT_ID}',
          ...EnvironmentMapping,
        },
        volumes: [
          {
            type: 'volume',
            source: 'chainfile',
            target: '/var/chainfile',
          },
        ],
        networks: {
          chainfile: {},
        },
      },
    };
  }

  private createContainers(): Record<string, object> {
    // TODO: resources (cpu, memory) is not supported yet for docker-compose
    //  https://docs.docker.com/compose/compose-file/compose-file-v3/#resources
    //  I'm not sure if we should since docker-compose typically runs on a single machine
    //  and utilizes the host's resources.
    //  Which limits its usefulness in orchestration when placement is not a concern.
    //  Adding them would make the compose hard limit the resources of the host even if the host
    //  has more resources available.

    function createPorts(container: Container): string[] {
      return Object.values(container.endpoints).map((endpoint) => {
        // TODO: Support Binding P2P Port Statically
        return `0:${endpoint.port}`;
      });
    }

    interface Volume {
      type: 'volume';
      source?: string;
      target: string;
    }

    function createVolumes(container: Container): Volume[] {
      const volumes: Volume[] = [
        {
          type: 'volume',
          source: 'chainfile',
          target: '/var/chainfile',
        },
      ];

      container.volumes?.persistent?.paths.forEach((path) => {
        volumes.push({
          type: 'volume',
          target: path,
        });
      });

      container.volumes?.ephemeral?.paths.forEach((path) => {
        volumes.push({
          type: 'volume',
          target: path,
        });
      });

      return volumes;
    }

    return Object.entries(this.definition.containers).reduce(
      (services, [name, container]) => {
        services[name] = {
          container_name: `${name}-$\{CHAINFILE_DEPLOYMENT_ID}`,
          image: container.image,
          command: container.command,
          environment: Object.entries(container.environment ?? {}).reduce(
            (env, [key, value]) => {
              return {
                ...env,
                [key]: typeof value === 'string' ? value : `$\{${value.key}}`,
              };
            },
            {} as Record<string, string>,
          ),
          ports: createPorts(container),
          volumes: createVolumes(container),
          networks: {
            chainfile: {},
          },
        };
        return services;
      },
      {} as Record<string, object>,
    );
  }
}
