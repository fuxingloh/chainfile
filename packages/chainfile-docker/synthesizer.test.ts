import { expect, it } from '@jest/globals';
import { Chainfile } from 'chainfile/schema';

import { Synthesizer } from './synthesizer';

it('should fail to synthesize with invalid chainfile', async () => {
  expect(() => new Synthesizer({})).toThrowError();
});

it('should synthesize with valid chainfile', async () => {
  const chainfile: Chainfile = {
    caip2: 'eip155:1337',
    name: 'Ganache',
    env: {
      RPC_USER: {
        type: 'RandomBytes',
        length: 16,
        encoding: 'hex',
      },
      RPC_PASSWORD: {
        type: 'RandomBytes',
        length: 16,
        encoding: 'hex',
      },
      URL: {
        type: 'Value',
        value: 'http://${RPC_USER}:${RPC_PASSWORD}@ganache:8554',
      },
    },
    containers: {
      ganache: {
        image: 'docker.io/trufflesuite/ganache@sha256:c62c58290c28e24b427f74c6f597ff696257bd2d8e8d517ce4cf46b29b304a3f',
        source: 'https://github.com/trufflesuite/ganache',
        environment: {
          RPCUSER: {
            key: 'RPC_USER',
          },
          RPCPASSWORD: {
            key: 'RPC_PASSWORD',
          },
        },
        // Orchestration
        resources: {
          cpu: 0.25,
          memory: 256,
        },
        // Endpoints
        endpoints: {
          p2p: {
            port: 8555,
          },
          rpc: {
            port: 8545,
            protocol: 'HTTP JSON-RPC 2.0',
            authorization: {
              type: 'HttpBasic',
              username: {
                key: 'RPC_USER',
              },
              password: {
                key: 'RPC_PASSWORD',
              },
            },
            probes: {
              readiness: {
                params: [],
                method: 'eth_blockNumber',
                match: {
                  result: {
                    type: 'string',
                  },
                },
              },
            },
          },
        },
        volumes: {
          persistent: {
            paths: ['/.ganache'],
            size: {
              initial: '1G',
              from: '2024-01-01',
              growth: '1G',
              rate: 'yearly',
            },
          },
        },
      },
    },
  };

  const synthesizer = new Synthesizer(chainfile);
  expect(synthesizer.synthCompose()).toMatchSnapshot();
  expect(synthesizer.synthEnv().split('\n')).toStrictEqual([
    expect.stringMatching(/^RPC_USER=[0-9a-f]{32}$/),
    expect.stringMatching(/^RPC_PASSWORD=[0-9a-f]{32}$/),
    expect.stringMatching(/^URL=http:\/\/[0-9a-f]{32}:[0-9a-f]{32}@ganache:8554$/),
    expect.stringMatching(/^CHAINFILE_DEPLOYMENT_ID=[0-9a-f]{16}$/),
  ]);
});

it('should have different deploymentId when using different Synthesizer', async () => {
  const env: Chainfile = {
    caip2: 'eip155:1337',
    name: 'Ganache',
    containers: {
      ganache: {
        image: 'docker.io/trufflesuite/ganache@sha256:c62c58290c28e24b427f74c6f597ff696257bd2d8e8d517ce4cf46b29b304a3f',
        source: 'https://github.com/trufflesuite/ganache',
        resources: {
          cpu: 0.25,
          memory: 256,
        },
        endpoints: {},
      },
    },
  };

  const synthesizer1 = new Synthesizer(env);
  const synthesizer2 = new Synthesizer(env);
  expect(synthesizer1.deploymentId).not.toEqual(synthesizer2.deploymentId);
});
