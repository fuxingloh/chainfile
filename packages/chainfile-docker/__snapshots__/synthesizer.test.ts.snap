// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should synthesize with valid chainfile 1`] = `
"# Generated by chainfile-docker, do not edit manually.
# Chainfile Version: 0.0.0
# Chainfile ID: eip155:1337/ganache:7.9.1
# Chainfile Name: Ganache
# Chainfile CAIP-2: eip155:1337

name: eip155_1337_ganache_7_9_1
services:
  chainfile-agent:
    container_name: chainfile-agent-\${CHAINFILE_DEPLOYMENT_ID}
    image: ghcr.io/vetumorg/chainfile-agent:0.0.0
    ports:
      - '0:1503'
    environment:
      CHAINFILE_JSON: >-
        {"id":"eip155:1337/ganache:7.9.1","caip2":"eip155:1337","name":"Ganache","env":{"RPC_USER":{"type":"RandomBytes","length":16,"encoding":"hex"},"RPC_PASSWORD":{"type":"RandomBytes","length":16,"encoding":"hex"},"URL":{"type":"Value","value":"http://$\${RPC_USER}:$\${RPC_PASSWORD}@ganache:8554"}},"containers":{"ganache":{"image":"docker.io/trufflesuite/ganache@sha256:c62c58290c28e24b427f74c6f597ff696257bd2d8e8d517ce4cf46b29b304a3f","source":"https://github.com/trufflesuite/ganache","environment":{"RPCUSER":{"key":"RPC_USER"},"RPCPASSWORD":{"key":"RPC_PASSWORD"}},"resources":{"cpu":0.25,"memory":256},"endpoints":{"p2p":{"port":8555},"rpc":{"port":8545,"protocol":"HTTP
        JSON-RPC
        2.0","authorization":{"type":"HttpBasic","username":{"key":"RPC_USER"},"password":{"key":"RPC_PASSWORD"}},"probes":{"readiness":{"params":[],"method":"eth_blockNumber","match":{"result":{"type":"string"}}}}}},"volumes":{"persistent":{"paths":["/.ganache"],"size":{"initial":"1G","from":"2024-01-01","growth":"1G","rate":"yearly"}}}}}}
      CHAINFILE_DEPLOYMENT_ID: \${CHAINFILE_DEPLOYMENT_ID}
      CHAINFILE_ENVIRONMENT_RPC_USER: \${RPC_USER}
      CHAINFILE_ENVIRONMENT_RPC_PASSWORD: \${RPC_PASSWORD}
      CHAINFILE_ENVIRONMENT_URL: \${URL}
    volumes:
      - type: volume
        source: chainfile
        target: /var/chainfile
    networks:
      chainfile: {}
  ganache:
    container_name: ganache-\${CHAINFILE_DEPLOYMENT_ID}
    image: docker.io/trufflesuite/ganache@sha256:c62c58290c28e24b427f74c6f597ff696257bd2d8e8d517ce4cf46b29b304a3f
    environment:
      RPCUSER: \${RPC_USER}
      RPCPASSWORD: \${RPC_PASSWORD}
    ports:
      - '0:8555'
      - '0:8545'
    volumes:
      - type: volume
        source: chainfile
        target: /var/chainfile
      - type: volume
        target: /.ganache
    networks:
      chainfile: {}
networks:
  chainfile: {}
volumes:
  chainfile: {}
"
`;
